### ISSUE TITLE
OptimisticLockException during TGT cleanup

### ISSUE THREAD

**Issue by PavlikPolivka** [apereo/cas#2658]
We are using CAS version  5.0.4 and we are running on 4 node cluster.
We are using JpaTicketRegistry with JpaLockingStrategy for locking the tickets during cleanup. 
The cleanup works fine on one node environment but on 4 nodes it is failing with OptimisticLockException.

```
javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [org.apereo.cas.ticket.TicketGrantingTicketImpl#****************************MqV4COEK0bmyEfe-usfr-sage-web1]
	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:197) ~[hibernate-core-5.2.2.Final.jar:5.2.2.Final]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:86) ~[hibernate-core-5.2.2.Final.jar:5.2.2.Final]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:174) ~[hibernate-core-5.2.2.Final.jar:5.2.2.Final]
	at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3392) ~[hibernate-core-5.2.2.Final.jar:5.2.2.Final]
	at org.hibernate.internal.SessionImpl.find(SessionImpl.java:3342) ~[hibernate-core-5.2.2.Final.jar:5.2.2.Final]
	at sun.reflect.GeneratedMethodAccessor291.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_45]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_45]
	at org.springframework.orm.jpa.SharedEntityManagerCreator$SharedEntityManagerInvocationHandler.invoke(SharedEntityManagerCreator.java:298) ~[spring-orm-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at com.sun.proxy.$Proxy177.find(Unknown Source) ~[?:?]
	at org.apereo.cas.ticket.registry.JpaTicketRegistry.getRawTicket(JpaTicketRegistry.java:96) ~[cas-server-support-jpa-ticket-registry-5.0.4.jar:5.0.4]
	at org.apereo.cas.ticket.registry.JpaTicketRegistry.getTicket(JpaTicketRegistry.java:82) ~[cas-server-support-jpa-ticket-registry-5.0.4.jar:5.0.4]
	at org.apereo.cas.ticket.registry.AbstractTicketRegistry.deleteTicket(AbstractTicketRegistry.java:104) ~[cas-server-core-tickets-5.0.4.jar:5.0.4]
	at org.apereo.cas.ticket.registry.AbstractTicketRegistry$$FastClassBySpringCGLIB$$d3c67a11.invoke(<generated>) ~[cas-server-core-tickets-5.0.4.jar:5.0.4]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:651) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.apereo.cas.ticket.registry.JpaTicketRegistry$$EnhancerBySpringCGLIB$$47e2ec3d.deleteTicket(<generated>) ~[cas-server-support-jpa-ticket-registry-5.0.4.jar:5.0.4]
	at sun.reflect.GeneratedMethodAccessor395.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_45]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_45]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:133) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:121) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213) ~[spring-aop-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at com.sun.proxy.$Proxy89.deleteTicket(Unknown Source) ~[?:?]
	at org.apereo.cas.ticket.registry.DefaultTicketRegistryCleaner.clean(DefaultTicketRegistryCleaner.java:75) ~[cas-server-core-tickets-5.0.4.jar:5.0.4]

```
This worked fine in CAS 4. Only major difference I can see in CAS 5 is thet the locks are now LockModeType.OPTIMISTIC and before it was PESSIMISTIC_WRITE.
I have no idea if changing it back to PESSIMISTIC_WRITE would help this because issue because I cannot easily extend the JpaLockingStrategy class because the Lock entity class is private static class now (changed in 5 as well). 

- CAS server version: 5.0.4
- Container/environment type/version (Tomcat, Jetty, Java, Linux/Windows, etc): Tomcat 8, Java 8, Centos

**Comment by dima767** [comment]
Which should be fine, no? 

The idea here is that a single resource (the database) is being used by multiple competing clients (CAS nodes) to perform the "clean up" transactions. If one node gets to it first, other nodes trying to act on the same piece of data will fail under optimistic locking scenario, and that is fine, IMHO. Just need to move on, as this exception is not "catastrophic" for cleaner worker... well, unless otherwise is demonstrated that this will prevent CAS server functioning properly...

**Comment by dima767** [comment]
... or just an idea ... in a multi-node deployment scenario with a single database for tickets storage, dedicate a single node to run the "cleaner" job, and turn the cleaner off on other nodes

**Comment by PavlikPolivka** [comment]
It's one exception per ticket. So it can be something like 25 exceptions per minute that are logged as ERROR. But yes it works and tickets are deleted.

Running the cleanup on just one node should be enough for us. Is there a similar option in 5.0.x?

**Comment by mmoayyed** [comment]
The option is the same.

**Comment by dima767** [comment]
Just want to make sure we are accurate here... The reason I said `since 5.1` is that I have not seen the similar code: https://github.com/apereo/cas/blob/master/core/cas-server-core-tickets/src/main/java/org/apereo/cas/config/CasCoreTicketsSchedulingConfiguration.java#L51-L59 in `5.0.x` branch

and this class: https://github.com/apereo/cas/blob/master/core/cas-server-core-tickets/src/main/java/org/apereo/cas/ticket/registry/NoOpTicketRegistryCleaner.java states that it's been added in `5.1.0`

**Comment by PavlikPolivka** [comment]
Found it here
https://github.com/apereo/cas/blob/5.0.x/core/cas-server-core-tickets/src/main/java/org/apereo/cas/ticket/registry/DefaultTicketRegistryCleaner.java#L44-L48

This seems to be working for me. It still calls the release on the lock, but that is fixed in 5.1.

Should I close this issue?

**Comment by dima767** [comment]
Ah, OK, I missed it ;-)

Yeah, I'll close the issue... But in general, I'd suggest you to look into upgrading to `5.1.0` which is a much more "solid" release ;-)
